tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,344 @@
+  // Arbitrary time for time-independent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>{});
+  const auto& psi = get<ScalarWave::Psi>(vars);
+  const auto& phi = get<ScalarWave::Phi<dim>>(vars);
+  const auto& pi = get<ScalarWave::Pi>(vars);
+  // Compute derivatives d_phi, d_pi, and d_gauge_function numerically
+  // First, prepare
+  using VariablesTags =
+      tmpl::list<ScalarWave::Psi, ScalarWave::Pi, ScalarWave::Phi<3>>;
+
+  Variables<VariablesTags> sw_vars(data_size);
+  get<ScalarWave::Psi>(sw_vars) = psi;
+  get<ScalarWave::Pi>(sw_vars) = pi;
+  get<ScalarWave::Phi<dim>>(sw_vars) = phi;
+  // Second, compute derivatives
+  const auto sw_derivs =
+      partial_derivatives<VariablesTags, VariablesTags, dim, Frame::Inertial>(
+          sw_vars, mesh, coord_map.inv_jacobian(x_logical));
+  const auto& deriv_psi =
+      get<Tags::deriv<ScalarWave::Psi, tmpl::size_t<dim>, Frame::Inertial>>(
+          sw_derivs);
+  const auto& deriv_pi =
+      get<Tags::deriv<ScalarWave::Pi, tmpl::size_t<dim>, Frame::Inertial>>(
+          sw_derivs);
+  const auto& deriv_phi = get<
+      Tags::deriv<ScalarWave::Phi<dim>, tmpl::size_t<dim>, Frame::Inertial>>(
+      sw_derivs);
+
+  // Insert into databox
+  const auto box = db::create<
+      db::AddSimpleTags<
+          ::Tags::Coordinates<dim, Frame::Inertial>, ScalarWave::Psi,
+          ScalarWave::Phi<dim>, ScalarWave::Pi,
+          ::Tags::deriv<ScalarWave::Psi, tmpl::size_t<dim>, Frame::Inertial>,
+          ::Tags::deriv<ScalarWave::Phi<dim>, tmpl::size_t<dim>,
+                        Frame::Inertial>,
+          ::Tags::deriv<ScalarWave::Pi, tmpl::size_t<dim>, Frame::Inertial>>,
+      db::AddComputeTags<ScalarWave::Tags::OneIndexConstraintCompute<dim>,
+                         ScalarWave::Tags::TwoIndexConstraintCompute<dim>>>(
+      x, psi, phi, pi, deriv_psi, deriv_phi, deriv_pi);
+
+  // Compute tested quantities locally
+  const auto one_index_constraint =
+      ScalarWave::one_index_constraint(deriv_psi, phi);
+  const auto two_index_constraint = ScalarWave::two_index_constraint(deriv_phi);
+
+  // Check that their compute items in databox furnish identical values
+  CHECK(db::get<ScalarWave::Tags::OneIndexConstraint<dim>>(box) ==
+        one_index_constraint);
+  CHECK(db::get<ScalarWave::Tags::TwoIndexConstraint<dim>>(box) ==
+        two_index_constraint);
+}
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.Constraints",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/ScalarWave/"};
+  INFO("Testing constraints with randomized input");
+  // Test the one-index constraint with random numbers
+  test_one_index_constraint_random<1>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
```

Comment:
Factor out the DataVector construction and name it `used_for_size`

Reply:
done